src/com/ms/silverking/cloud/dht/client/impl/SynchronousNamespacePerspectiveImpl.java

```diff
@@ -19,158 +19,161 @@
+    AsyncValueRetrieval<K, V> asyncRetrieval;
+
+    asyncRetrieval = (AsyncValueRetrieval<K, V>) baseRetrieve(keys, getOptions, opLWTMode);
+    asyncRetrieval.waitForCompletion();
+    return asyncRetrieval.getValues();
+  }
+
+  @Override
+  public Map<K, V> get(Set<? extends K> keys) throws RetrievalException {
+    return get(keys, nspoImpl.getDefaultGetOptions());
+  }
+
+  @Override
+  public V get(K key, GetOptions getOptions) throws RetrievalException {
+    StoredValue<V> storedValue;
+
+    storedValue = retrieve(key, getOptions);
+    if (storedValue != null) {
+      return storedValue.getValue();
+    } else {
+      return null;
     }
-    
-    public void invalidate(K key, InvalidationOptions invalidationOptions) throws InvalidationException {
-        invalidate(ImmutableSet.of(key), invalidationOptions);
+  }
+
+  @Override
+  public V get(K key) throws RetrievalException {
+    return get(key, nspoImpl.getDefaultGetOptions());
+  }
+
+  @Override
+  public Map<K, V> waitFor(Set<? extends K> keys, WaitOptions waitOptions) throws RetrievalException {
+    AsyncValueRetrieval<K, V> asyncRetrieval;
+
+    asyncRetrieval = (AsyncValueRetrieval<K, V>) baseRetrieve(keys, waitOptions, opLWTMode);
+    asyncRetrieval.waitForCompletion();
+    return asyncRetrieval.getValues();
+  }
+
+  @Override
+  public Map<K, V> waitFor(Set<? extends K> keys) throws RetrievalException {
+    return waitFor(keys, nspoImpl.getDefaultWaitOptions());
+  }
+
+  @Override
+  public V waitFor(K key, WaitOptions waitOptions) throws RetrievalException {
+    StoredValue<V> storedValue;
+
+    storedValue = retrieve(key, waitOptions);
+    if (storedValue != null) {
+      return storedValue.getValue();
+    } else {
+      return null;
     }
-    
-    public void invalidate(K key) throws InvalidationException {
-        invalidate(key, nspoImpl.getDefaultInvalidationOptions());
+  }
+
+  @Override
+  public V waitFor(K key) throws RetrievalException {
+    return waitFor(key, nspoImpl.getDefaultWaitOptions());
+  }
+
+  // writes
+
+  @Override
+  public void put(Map<? extends K, ? extends V> values, PutOptions putOptions) throws PutException {
+    basePut(values, putOptions, opLWTMode).waitForCompletion();
+  }
+
+  @Override
+  public void put(Map<? extends K, ? extends V> values) throws PutException {
+    put(values, nspoImpl.getDefaultPutOptions());
+  }
+
+  @Override
+  public void put(K key, V value, PutOptions putOptions) throws PutException {
+    put(ImmutableMap.of(key, value), putOptions);
+  }
+
+  @Override
+  public void put(K key, V value) throws PutException {
+    put(key, value, nspoImpl.getDefaultPutOptions());
+  }
+
+  public void invalidate(Set<? extends K> keys, InvalidationOptions invalidationOptions) throws InvalidationException {
+    try {
+      baseInvalidation(keys, invalidationOptions, nspoImpl.getValueSerializer(), opLWTMode).waitForCompletion();
+    } catch (PutException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)